services/http.go

```diff
@@ -0,0 +1,104 @@
+package services
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/getlantern/flashlight/v7/common"
+)
+
+const (
+	// retryWaitSeconds is the base wait time in seconds between retries
+	retryWaitSeconds = 5 * time.Second
+	maxRetryWait     = 10 * time.Minute
+)
+
+// sender is a helper for sending post requests. If the request fails, sender calulates an
+// exponential backoff time using retryWaitSeconds and return it as the sleep time.
+type sender struct {
+	failCount      int
+	atMaxRetryWait bool
+}
+
+// post posts data to the specified URL and returns the response, the sleep time in seconds, and any
+// error that occurred.
+//
+// Note: if the request is successful, it is the responsibility of the caller to read the response
+// body to completion and close it.
+func (s *sender) post(
+	originURL string,
+	buf io.Reader,
+	rt http.RoundTripper,
+	user common.UserConfig,
+) (*http.Response, int64, error) {
+	resp, err := s.doPost(originURL, buf, rt, user)
+	if err == nil {
+		if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusNoContent {
+			err = fmt.Errorf("bad response code: %v", resp.StatusCode)
+			goto backoff
```

Comment:
This is a optional suggestion since it's comprehensible and not complex case use of `goto`: We could replace this `goto` creating and calling a `retry` function (I'm assuming it's  going to be retried later), something like this:
```
func (s *sender) retry(err error) (*http.Response, int64, error) {
	if s.atMaxRetryWait {
		// we've already reached the max wait time, so we don't need to perform the calculation again.
		// we'll still increment the fail count to keep track of the number of failures
		s.failCount++
		return nil, int64(maxRetryWait.Seconds()), err
	}

	wait := time.Duration(math.Pow(2, float64(s.failCount))) * retryWaitSeconds
	s.failCount++

	if wait > maxRetryWait {
		s.atMaxRetryWait = true
		return nil, int64(maxRetryWait.Seconds()), err
	}

	return nil, int64(wait.Seconds()), err
}
```
And replace line 42 and 58-74 by `s.retry(err)`. We could also use a library to handle retries like https://github.com/avast/retry-go or https://github.com/sethvargo/go-retry

Reply:
> This is a optional suggestion since it's comprehensible and not complex case use of goto

Yeah, I need to change that. 

> We could also use a library to handle retries like https://github.com/avast/retry-go or https://github.com/sethvargo/go-retry

I prefer not to use external libraries for small, simple things like this. There's less dependencies.